lib/System/IO/Utf8.hs

```diff
@@ -0,0 +1,99 @@
+{- SPDX-FileCopyrightText: 2020 Serokell <https://serokell.io/>
+ -
+ - SPDX-License-Identifier: MPL-2.0
+ -}
+
+{-# LANGUAGE LambdaCase   #-}
+{-# LANGUAGE ViewPatterns #-}
+
+-- | System IO for the modern world.
+--
+-- Standard IO functions assume that the character encoding of the data
+-- they read or write is the same as the one used by current locale. In many
+-- situtations this assumption is wrong, as tools work with files, and
+-- the files nowadays are mostly UTF-8 encoded, regardless of the locale.
+-- Therefore, it is almost always a good idea to switch the encoding of
+-- file handles to UTF-8.
+--
+-- The same applies to standard input, output, and error handles. However,
+-- there is an edge-case: if they are attached to a terminal, and the
+-- encoding is not UTF-8, using UTF-8 might actually be unsafe.
+--
+-- Functions in this module help deal with all these issues.
+module System.IO.Utf8
+  ( withUtf8StdHandles
+
+  , hSetEncoding
+
+  , openFile
+  , withFile
+  ) where
+
+import Control.Exception.Safe (MonadMask, bracket)
+import Control.Monad.IO.Class (MonadIO, liftIO)
+import GHC.IO.Encoding (mkTextEncoding, textEncodingName, utf8)
+import System.IO (stderr, stdin, stdout)
+
+import qualified System.IO as IO
+
+
+type EncRestoreAction = IO.Handle -> IO ()
```

Comment:
Is there a use case when we set the best encoding for one handler, but apply returned `EpcRestoreAction` to another handler?

Reply:
I don’t think so, but the idea is that this action is return from an “acquirer” of `bracket`, so it’s just easier to write it this way. This way we rely on `bracket` to tell us which resource to release, and it feels more natural and safe.

Reply:
That sounds convenient. Though, why not make it just `IO ()` then?

Reply:
Because we want to get the handle we will run on from `bracket`, instead of remembering the one we were passed initially. I’m pretty sure these are always the same, but I fell safer getting it from `bracket`.